### PR TITLE
APPDEV-1170 Crashlytics crash report in apputils class

### PR DESCRIPTION
**Pull request by Sivateja-CMI**
-- Added support for showing user Toast when exception occurs in non-main threads.

### CODE REVIEW

File: app/src/main/java/nu/yona/app/utils/AppUtils.java [hunk 1]
```diff
@@ -288,25 +289,37 @@ public static String getHourDigit(String digit)
...
+			logExceptionToCrashlytics(e);
 		}
-		if (listener != null)
+	}
+
+	/**
+	 * getErrorMessage from give Exception or return a generic error message
+	 */
+
+	public static ErrorMessage getErrorMessage(Exception e)
```
Comment (Bert-R) [inline app/src/main/java/nu/yona/app/utils/AppUtils.java]: This method doesn't have to be public.

Comment (Bert-R) [inline app/src/main/java/nu/yona/app/utils/AppUtils.java]: Generally don't use such short names. Exceptions to that rule are: lambda expressions, loop iterator value,  name of exception in catch block

File: app/src/main/java/nu/yona/app/utils/AppUtils.java [hunk 2]
```diff
@@ -288,25 +289,37 @@ public static String getHourDigit(String digit)
...
+	{
+		ErrorMessage errorMessage = new ErrorMessage();
+		if (e != null && e.getMessage() != null)
 		{
-			listener.onError(errorMessage);
+			errorMessage.setMessage(e.getMessage());
+			Logger.loge(e.getClass().getSimpleName(), e.getMessage());
```
Comment (Bert-R) [inline app/src/main/java/nu/yona/app/utils/AppUtils.java]: This shouldn't be part of ``getErrorMessage``. A ``get`` method should just get something.
Move the logging to ``reportException``

File: app/src/main/java/nu/yona/app/utils/AppUtils.java [hunk 3]
```diff
@@ -288,25 +289,37 @@ public static String getHourDigit(String digit)
...
 		}
 		else
 		{
-			showErrorToast(errorMessage);
-			logExceptionToCrashlytics(e);
+			String message = YonaApplication.getAppContext().getString(R.string.generic_error_message);
+			errorMessage.setMessage(message);
+			Logger.loge(e.getClass().getSimpleName(), message);
```
Comment (Bert-R) [inline app/src/main/java/nu/yona/app/utils/AppUtils.java]: This line is nearly identical to the one I just commented on. It could easily be put below the  if/else and take the message text from the just constructed error message. Now it should be removed, as said in the previous comment.

File: app/src/main/java/nu/yona/app/utils/AppUtils.java [hunk 4]
```diff
@@ -288,25 +289,37 @@ public static String getHourDigit(String digit)
...
 		}
-		Logger.loge(e.getClass().getSimpleName(), e.getMessage());
+		return errorMessage;
```
Comment (Bert-R) [inline app/src/main/java/nu/yona/app/utils/AppUtils.java]: Please make the method this simple:
```
	public static ErrorMessage getErrorMessage(Exception e)
	{
		ErrorMessage errorMessage = new ErrorMessage();
		if (e != null && e.getMessage() != null)
		{
			return new ErrorMessage(e.getMessage());
		}
		return new ErrorMessage(YonaApplication.getAppContext().getString(R.string.generic_error_message));
	}
```

Please also add ``ErrorMessage`` to your refactoring list: this is a perfect class for immutable behavior. Just remove all setters and only allow taking the values into the constructor.

File: app/src/main/res/values-nl/strings.xml [hunk 5]
```diff
@@ -222,5 +222,6 @@
     <string name="generic_alert_title">Waarschuwing</string>
     <string name="generic_exception_message">Onbekende fout. Neem contact op met Yona support.</string>
     <string name="generic_mobile_number_format_error">Ongeldig telefoonnummer. Controleer het formaat</string>
+    <string name="generic_error_message">Onbekende fout. Neem contact op met Yona support.</string>
```
Comment (Bert-R) [inline app/src/main/res/values-nl/strings.xml]: This message already exists on line 223.

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -224,4 +224,5 @@
     <string name="generic_alert_title">Alert</string>
     <string name="generic_exception_message">Unknown Exception. Please contact Yona support</string>
     <string name="generic_mobile_number_format_error">Invalid Mobile Number. Please check the format</string>
+    <string name="generic_error_message">Unknown Error. Please contact Yona support</string>
```
Comment (Bert-R) [inline app/src/main/res/values/strings.xml]: This message already exists on line 225.

File: app/src/main/java/nu/yona/app/api/model/ErrorMessage.java [hunk 7]
```diff
@@ -74,15 +78,6 @@ public String getCode()
 		return code;
 	}
 
-	/**
-	 * Sets code.
-	 *
-	 * @param code The code
-	 */
-	public void setCode(String code)
```
Comment (Bert-R) [inline app/src/main/java/nu/yona/app/api/model/ErrorMessage.java]: With the setters gone now, please make the members ``final``.

File: app/src/main/java/nu/yona/app/utils/AppUtils.java [hunk 8]
```diff
@@ -282,53 +283,58 @@ public static String getHourDigit(String digit)
...
 		}
-		Logger.loge(e.getClass().getSimpleName(), e.getMessage());
+		Logger.loge(className, errorMessage.getMessage());
+	}
+
+	/**
+	 * getErrorMessage from give Exception or return a generic error message
+	 */
+
+	public static ErrorMessage getErrorMessageFromException(Exception exception)
```
Comment (Bert-R) [inline app/src/main/java/nu/yona/app/utils/AppUtils.java]: Make this one ``private``.

File: app/src/test/java/nu/yona/app/api/manager/impl/AuthenticateManagerImplTest.java [hunk 9]
```diff
@@ -34,211 +31,240 @@
...
-                ArgumentMatchers.any(DataLoadListener.class));
-
-        Mockito.when(authenticateDAOMock.getUser()).thenReturn(getMockedUser());
-    }
-
-    @After
-    public void tearDown() {
-        dbhelper.close();
-    }
+public class AuthenticateManagerImplTest extends YonaTestCase
```
Comment (Bert-R) [inline app/src/test/java/nu/yona/app/api/manager/impl/AuthenticateManagerImplTest.java]: I see I didn't reformat the tests. Sorry about that. Can you do a separate PR with a reformat of the remaining tests?

Comment (Sivateja-CMI) [reply app/src/test/java/nu/yona/app/api/manager/impl/AuthenticateManagerImplTest.java]: ok. will do that
